### ISSUE TITLE
On Python 3.13, "greenlet.error: cannot switch to a different thread"

### ISSUE THREAD

**Issue by itamarst** [python-greenlet/greenlet#425]
Eventlet runs on greenlet, but on Python 3.13rc2 tests that work on older versions of Python now fail (using this PR: https://github.com/eventlet/eventlet/pull/966, and greenlet 3.1.0):

```
$ .tox/py313/bin/python tests/isolated/patcher_threading_join.py 
Traceback (most recent call last):
  File "/home/itamarst/devel/eventlet/eventlet/hubs/hub.py", line 471, in fire_timers
    timer()
    ~~~~~^^
  File "/home/itamarst/devel/eventlet/eventlet/hubs/timer.py", line 59, in __call__
    cb(*args, **kw)
    ~~^^^^^^^^^^^^^
  File "/home/itamarst/devel/eventlet/eventlet/semaphore.py", line 147, in _do_acquire
    waiter.switch()
    ~~~~~~~~~~~~~^^
greenlet.error: cannot switch to a different thread
```

It's not clear if this is greenlet's fault or if eventlet is doing something wrong (but something wrong that used to work!). If you have any hints on how to debug this sort of problem that would be helpful in trying to fix it in eventlet.

**Comment by jamadden** [comment]
Hmm. The [patch to add 3.13 support](https://github.com/python-greenlet/greenlet/commit/528b81680cc95a86ab25370ea3f0461384b6d753) didn't touch the code that [detects when a greenlet is switching threads](https://github.com/python-greenlet/greenlet/blob/db5c781da4be56e3fe28f183c2832bbefd15bb5f/src/greenlet/TGreenlet.cpp#L217).

I modified the test case to print the actual thread ident and native thread id, both before spawning (from the main greenlet/main thread), and from inside the spawned thread before it sleeps; they also print the current greenlet. I also modified the code that's about to throw `greenlet.error` to print what the current thread's main greenlet is, and what the owning main greenlet is (the main greenlet from the thread that spawned the greenlet):

```
spawning in thread_ident=0x1ea8bcf40; native_id=0xe511de
spawning in main greenlet= <greenlet.greenlet object at 0x102a80d70 (otid=0x102a813a0) current active started main>

About to sleep; thread_ident=0x16ea33000; native_id=0xe511e5
Sleeping in greenlet <greenlet.greenlet object at 0x102c30e10 (otid=0x102c30a40) current active started main>

About to raise greenlet.error
Current Main Greenlet = <greenlet.greenlet object at 0x102c30e10 (otid=0x102c30a40) suspended active started main>
Owning Main Greenlet = <greenlet.greenlet object at 0x102a80d70 (otid=0x102a813a0) suspended active started main>
```

The original thread IDs are not the same as the thread IDs in the new Thread object. Nor is the main greenlet the same in either case. 

The greenlet error code is correctly identifying both the current greenlet and the owning greenlet (they match what the Python code printed).

Thus, I believe the patching of `threading._start_joinable_thread` is incorrect. Indeed, printing the value of that *after* the monkey-patching reveals it to be the version from the `_thread` module. If I had to guess, I would guess this is because `threading` is imported earlier than it used to be; it appears that the eventlet code is only patching `_thread.start_joinable_thread` and not patching `threading._start_joinable_thread`. So if `threading` is imported before patching, you get native threads.
